File: src/main/java/com/iotassistant/models/rules/OutOfRangeGpsRule.java
package com.iotassistant.models.rules;

import javax.persistence.Entity;

import com.iotassistant.models.devices.GpsPosition;
import com.iotassistant.models.notifications.NotificationTypeEnum;

@Entity
public class OutOfRangeGpsRule extends GpsRule{
	
	private String latitude;
	
	private String longitude;
	
	private int range;
	
	public OutOfRangeGpsRule() {
		super();
	}

	public OutOfRangeGpsRule(String gpsName, boolean enabled, RuleTriggerIntervalEnum timeBetweenTriggers,
			NotificationTypeEnum notificationType, String latitude, String longitude, int range) {
		super(gpsName, enabled, timeBetweenTriggers, notificationType);
		this.latitude = latitude;
		this.longitude = longitude;
		this.range = range;
	}

	public String getLatitude() {
		return latitude;
	}

	public String getLongitude() {
		return longitude;
	}

	public int getRange() {
		return range;
	}

	@Override
	public void accept(GpsRuleVisitor gpsRuleVisitor) {
		gpsRuleVisitor.visit(this);	
	}

	@Override
	public boolean apply(String gpsName, GpsPosition position) {
		return this.getGpsName().equals(gpsName) && calculateDistanceInMeters(position) >= range;
	}
	
	public double calculateDistanceInMeters(GpsPosition position) {
		return org.apache.lucene.util.SloppyMath.haversinMeters(
				Float.parseFloat(this.latitude), 
				Float.parseFloat(this.longitude), 
				Float.parseFloat(position.getLatitude()), 
				Float.parseFloat(position.getLongitude()));
}


}

File: src/main/resources/static/app/rules/gpsrules/entities/OutOfRangeGpsRule.js
class OutOfRangeGpsRule extends GpsRule {
	
	constructor(latitude, longitude, range, gpsName, id, enabled, timeBetweenTriggers, notificationType) {
		super(gpsName, id, enabled, timeBetweenTriggers, notificationType);
		this.gpsName = gpsName;
		this.latitude = latitude;
		this.longitude = longitude;
		this.range = range;
	}
	
	getLatitude() {
		return this.latitude;
	}
	
	getLongitude() {
		return this.longitude;
	}
	
	getRange() {
		return this.range;
	}
	
	isValid() {
		return super.isValid() &&  this.positionIsValid() && 0 < this.range;
	}
	
	positionIsValid() {
		return this.longitude !== 'undefined' && -180 <= this.longitude && this.longitude <= 180 
		&& this.latitude !== 'undefined' && -90 <= this.latitude && this.latitude <= 90;
	}
	
	static get typeString() {
    	return 'Out Of Range';
  	}
}